core/base/v7/inc/ROOT/TDirectoryEntry.h

```diff
@@ -26,77 +26,70 @@ namespace Experimental {
+  explicit TDirectoryEntry(const std::shared_ptr<T>& ptr):
+    fTypeInfo(&typeid(T)),
+    fType(TClass::GetClass(typeid(T))),
+    fObj(ptr) {}
 
   /// Get the last change date of the entry.
   const time_point_t& GetDate() const { return fDate; }
 
-  /// Get the object's type.
-  TClass* GetType() const { return fType; }
-
   /// Inform the entry that it has been modified, and needs to update its
   /// last-changed time stamp.
   void SetChanged() { fDate = clock_t::now(); }
 
-  /// Abstract interface to retrieve the address of the object represented by
-  /// this entry, if any.
-  virtual void* GetObjectAddr() const = 0;
-
-  /// Abstract interface to retrieve the type of the object represented by this
-  /// entry.
-  virtual const std::type_info& GetTypeInfo() const = 0;
-};
-
-
-/**
- Type-aware part of an entry in a TDirectory. It can manage the lifetime of an
- object, or only inform about last-change date and the type of an object.
- */
-
-template <class T>
-class TDirectoryEntryPtr: public TDirectoryEntryPtrBase {
-public:
-  /// Initialize a TDirectoryEntryPtr from an existing object ("write").
-  TDirectoryEntryPtr(T&& obj):
-    TDirectoryEntryPtrBase(TClass::GetClass(typeid(T))),  fObj(obj) {}
-
-  /// Initialize a TDirectoryEntryPtr from an existing object ("write").
-  TDirectoryEntryPtr(const std::shared_ptr<T>& ptr): fObj(ptr) {}
+  /// Type of the object represented by this entry.
+  const std::type_info& GetTypeInfo() const { return *fTypeInfo; }
 
-  virtual ~TDirectoryEntryPtr() {}
+  /// Get the object's type.
+  TClass* GetType() const { return fType; }
 
   /// Retrieve the `shared_ptr` of the referenced object.
-  std::shared_ptr<T>& GetPointer() const { return fObj; }
+  std::shared_ptr<void>& GetPointer() { return fObj; }
+  const std::shared_ptr<void>& GetPointer() const { return fObj; }
 
-  /// Retrieve the address of the object managed by this entry.
-  /// Returns `nullptr` if no object is managed by this entry.
-  void* GetObjectAddr() const final { return fObj.get(); }
+  template<class U>
+  std::shared_ptr<U> CastPointer() const;
 
-  /// Type of the object represented by this entry.
-  const std::type_info& GetTypeInfo() const final { return typeid(T); }
+  explicit operator bool() const { return *fTypeInfo != typeid(std::nullptr_t); }
 
 private:
-  std::shared_ptr<T> fObj; ///< Object referenced by this entry, if any.
+  time_point_t fDate = clock_t::now(); ///< Time of last change
+  const std::type_info* fTypeInfo;
+  TClass* fType;
+  std::shared_ptr<void> fObj;
 };
 
+template<class U>
+std::shared_ptr<U> TDirectoryEntry::CastPointer() const
+{
+  if (auto ptr = fType->DynamicCast(TClass::GetClass(typeid(U)), fObj.get()))
+    return std::shared_ptr<U>(fObj, static_cast<U*>(ptr));
+  return std::shared_ptr<U>();
+}
+
+inline bool operator==(const TDirectoryEntry& lhs, const TDirectoryEntry& rhs)
+{
+  return lhs.GetPointer() == rhs.GetPointer()
+    && lhs.GetTypeInfo() == rhs.GetTypeInfo();
```

Comment:
Well, there are some inconsistencies here. `operator==()` was added, AFAIR, only for `operator bool()`¸ to compare with empty `TDirectoryEntry`. I'm not sure whether we need `operator()==` at all. And it was added before I've reintroduced timestamps to `TDirectoryEntry`. Timestamps, in turn, were added just because they were in the original implementation. I don't understand how they will be updated on object changes, so their presence is itself questionable.